src/main/java/fi/aalto/cs/apluscourses/utils/async/TaskManager.java

```diff
@@ -0,0 +1,57 @@
+package fi.aalto.cs.apluscourses.utils.async;
+
+import java.util.List;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * An abstract interface for asynchronous execution following the fork-join model of task
+ * parallelism.
+ *
+ * <p>Note that it is not, however, required that the implementing classes actually use any
+ * concurrent computation.</p>
+ *
+ * @param <T> Type of the task objects (or futures) that are used as identifiers for asynchronous
+ *            tasks
+ */
+public interface TaskManager<T> {
```

Comment:
I have certain doubts, that we actually need our own Fork/Join ~~with blackjack and ...~~ for this task. If the idea is to parallelize some processes like installation and modules fetching, then maybe the [parallelStream()](https://docs.oracle.com/javase/8/docs/api/java/util/Collection.html#parallelStream--) will do? 
What comes to the asynchrony, then a direct usage of `CompletableFuture` looks fit here.

I don't have much to say about the implementation - it looks cool, but I'm not sure if the tasks we work with actually require this level of complexity and the benefits gained from introducing such conceptual and maintenance load overweight the losses.

Curious reading [comparing Fork/Join Framework vs. Parallel Streams vs. ExecutorService](https://blog.overops.com/forkjoin-framework-vs-parallel-streams-vs-executorservice-the-ultimate-benchmark/) found on the internet.

Reply:
My point here was to postpone the design decision which threading framework and thread pool we use by abstracting the task parallelism behind a single interface, which can easily be implemented with `CompleteableFuture`s, `Thread`s or, I'd guess, any task based parallel library, and allow execution to be dynamically changed to single-threaded for debugging and testing purposes.

Perhaps, when we are sure which threading framework we're gonna use and how, we can remove this extra layer of abstraction?